internal/oci/runtime_oci.go

```diff
@@ -956,11 +956,16 @@ killContainer:
 	// We cannot use ExponentialBackoff() here as its stop conditions are not flexible enough.
 	kwait.BackoffUntil(func() {
 		if _, err := r.runtimeCmd("kill", c.ID(), "KILL"); err != nil {
+			if errors.Is(err, unix.ESRCH) {
+				// Process is already gone, stop the loop.
+				stop()
+				return
+			}
```

Comment:
The current value of errno is not passed between different processes. This is not the error CRI-O sees.

Reply:
My previous approach was to check for the string ` if strings.Contains(err.Error(), "No such process")` but it doesn't appear like a full-proof one.

Reply:
> The current value of errno is not passed between different processes. This is not the error CRI-O sees.

whoops, I think that's my bad.

Reply:
This is not great, but for now this probably would have to do. Albeit, abstracting this away into a helper function would be an option, too.

A better option would be to talk to @kolyshkin and ask for more detailed exit codes to runc and crun so CRI-O could attempt to differentiate what happened. Currently, the exit code is just **1** most of the time on error.

Or, the runtime could also provide JSON document to stdout, as it does for other things such as `start` or `status` when terminating with a non-zero exit code, where said the document would carry a failure message and some other useful details we could parse on our side. However, I am not sure if this would be OCI-compliant. But probably on an error we could not care as much.

Reply:
This is what happens at the moment:

```console
([]string) (len=3 cap=3) {
 (string) (len=4) "kill",
 (string) (len=64) "d2b22ef4e5909c375e834aadda5fe7e16b74044e638ef164fd4625e95023264c",
 (string) (len=4) "KILL"
}
([]string) (len=6 cap=8) {
 (string) (len=6) "--root",
 (string) (len=9) "/run/runc",
 (string) (len=16) "--systemd-cgroup",
 (string) (len=4) "kill",
 (string) (len=64) "d2b22ef4e5909c375e834aadda5fe7e16b74044e638ef164fd4625e95023264c",
 (string) (len=4) "KILL"
}
(bytes.Buffer) open pidfd: No such process     <--- stderr (contains newline)

(bytes.Buffer)                                                          <--- stdout (empty)
(*exec.ExitError)(0xc00048be60)(exit status 1) <--- crun's exit code
```

This is what we then see as:

```console
ERRO[2024-07-31 14:19:55.694146382Z] Killing container d2b22ef4e5909c375e834aadda5fe7e16b74044e638ef164fd4625e95023264c failed: `/usr/bin/crio-crun --root /run/runc --systemd-cgroup kill d2b22ef4e5909c375e834aadda5fe7e16b74044e638ef164fd4625e95023264c KILL` failed: open pidfd: No such process
 : exit status 1
```

Reply:
@kwilczynski this needs to be done in runtime-spec first, so that multiple runtimes (runc, crun, youki) can behave in the same way. I think the best way to start the discusion about it is to open a PR against https://github.com/opencontainers/runtime-spec.

Reply:
cc @giuseppe ^^^

Reply:
@kolyshkin, sounds good! I will get to it then.

Reply:
@sohankunkerkar, I will take it to the runtime-spec community for feedback.

Meanwhile, we can proceed with a non-zero exit code plus a stderr string match. I suggested a helper function to do since we need to cover runc and crun, and they might have slightly different messages per:

- runc

```console
ERRO[2024-08-01 15:18:24.576675123Z] Killing container 1b3e518083c85f13bbe9b145da9f4f0dc7c254e48ce9fd7decbbcd6ae46acc1e failed: `/usr/bin/crio-runc --root /run/runc --systemd-cgroup kill 1b3e518083c85f13bbe9b145da9f4f0dc7c254e48ce9fd7decbbcd6ae46acc1e KILL` failed: time="2024-08-01T15:18:24Z" level=error msg="container not running"
 : exit status 1 
```

- crun

```console
 ERRO[2024-08-01 15:29:41.701878085Z] Killing container 52a27641c62eb42ec810d689c637dc86a329cdae2d46880c83a47122306d17f1 failed: `/usr/bin/crio-crun --root /run/runc --systemd-cgroup kill 52a27641c62eb42ec810d689c637dc86a329cdae2d46880c83a47122306d17f1 KILL` failed: open pidfd: No such process
 : exit status 1 
```

Also, runc can also emit an "invalid process" error, at some circumstances, I think.